File: 1207 Unique Number of Occurrences/1207unique-number-of-occurrences.cpp
class Solution {
public:
    bool uniqueOccurrences(vector<int>& arr) {
        sort(arr.begin(),arr.end());
        vector<int> v1;
        int count = 1;
        int n = arr.size();
        if(n==1 || n==2 && arr[0] == arr[1]){
            return true;
        }else if(n==2 && arr[0]!=arr[1]){
            return false;
        }else{
            for(int i=1;i<n;i++){
                if(arr[i] == arr[i-1]){
                    count++;
                }else{
                    v1.push_back(count);
                    count = 1;
                }
                if(i == n-1){
                    v1.push_back(count);
                }
            }
        }
        sort(v1.begin(),v1.end());
        int m = v1.size();
        // for(int i=0;i<m;i++){
        //     cout<<v1[i];
        // }
        for(int i=1;i<m;i++){
            if(v1[i] == v1[i-1]){
                return false;
            }
        }
        return true;
    }
};

File: 1207-unique-number-of-occurrences/1207-unique-number-of-occurrences.py
class Solution:
    def uniqueOccurrences(self, arr: List[int]) -> bool:

        # '''
        # Pattern - 

        # TC - O(n)
        # SC - O(1)
        # '''

        # f_arr = [[0,0] for _ in range(1001)]

        # o_map = defaultdict(int)

        # for num in arr:
        #     if num < 0:
        #         f_arr[-num][1] += 1
        #     else:
        #         f_arr[num][0] += 1
        
        # for a,b in f_arr:            
        #     if a>0:
        #         o_map[a] += 1
            
        #     if b>0:
        #         o_map[b] += 1
            
        #     if o_map[a]>1 or o_map[b]>1:
        #         return False
        
        # return True


        '''
        Pattern - 

        TC - O(n)
        SC - O(1)
        '''

        f_arr = [[0,0] for _ in range(1001)]

        o_set = set()

        for num in arr:
            if num < 0:
                f_arr[-num][1] += 1
            else:
                f_arr[num][0] += 1
        
        for a,b in f_arr:            
            if a>0:
                if a in o_set:
                    return False
                o_set.add(a)
            
            if b>0:
                if b in o_set:
                    return False
                o_set.add(b)
                    
        return True
        

